BacktraceScreen.c

```diff
@@ -0,0 +1,384 @@
+
+#include "BacktraceScreen.h"
+
+#if defined(HAVE_BACKTRACE)
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/wait.h>
+
+#include "CRT.h"
+#include "FunctionBar.h"
+#include "ListItem.h"
+#include "Macros.h"
+#include "Object.h"
+#include "Panel.h"
+#include "Platform.h"
+#include "Process.h"
+#include "ProvideCurses.h"
+#include "RichString.h"
+#include "Settings.h"
+#include "Vector.h"
+#include "XUtils.h"
+
+
+#define BACKTRACE_FRAME_HEADER_INDEX -1
+
+#define BACKTRACE_PANEL_OPTIONS_NAME_DEMANGLE "Demangle"
+#define BACKTRACE_PANEL_OPTIONS_NAME_RAW "Raw"
+
+#define BACKTRACE_PANEL_OPTIONS_OBJECT_PATH_FULL "Full Path"
+#define BACKTRACE_PANEL_OPTIONS_OBJECT_PATH_BASENAME "Basename"
+
+#define BACTRACE_PANEL_HEADER_NUMBER_FRAME "#"
+#define BACTRACE_PANEL_HEADER_ADDRESS "ADDRESS"
+#define BACTRACE_PANEL_HEADER_NAME "NAME"
+#define BACTRACE_PANEL_HEADER_PATH "PATH"
+
+#define MAX_ADDR_32 (0xffffffff)
+#define MAX_HEX_ADDR_STR_LEN_32 (strlen("0x") + 8)
+#define MAX_HEX_ADDR_STR_LEN_64 (strlen("0x") + 16)
+
+static const char* const BacktraceScreenFunctions[] = {
+   "Refresh",
+#if defined(HAVE_DEMANGLING)
+   BACKTRACE_PANEL_OPTIONS_NAME_RAW,
+#endif
+   BACKTRACE_PANEL_OPTIONS_OBJECT_PATH_FULL,
+   "Done   ",
+   NULL
+};
+
+static const char* const BacktraceScreenKeys[] = {
+   "F1",
+#if defined(HAVE_DEMANGLING)
+   "F2",
+#endif
+   "F3",
+   "Esc",
+   NULL
+};
+
+static const int BacktraceScreenEvents[] = {
+   KEY_F(1),
+#if defined(HAVE_DEMANGLING)
+   KEY_F(2),
+#endif
+   KEY_F(3),
+   27,
+};
+
+static void BacktraceFrame_displayHeader(const BacktraceFrame* frame, RichString* out) {
+   const BacktracePanelPrintingHelper* printingHelper = &frame->backtracePanel->printingHelper;
+   assert(printingHelper);
+
+   const BacktraceScreenDisplayOptions displayOptions = frame->backtracePanel->displayOptions;
+
+   int maxFunctionNameLength = printingHelper->maxFuncNameLen;
+   if (!!(displayOptions & DEMANGLE_NAME_FUNCTION) &&
+         printingHelper->maxDemangledFuncNameLen > 0) {
+      maxFunctionNameLength = printingHelper->maxDemangledFuncNameLen;
+   }
+
+   char* line = NULL;
+   int len = xAsprintf(&line, "%-*s %-*s %-*s %-*s",
+      (int)printingHelper->maxFrameNumLen, BACTRACE_PANEL_HEADER_NUMBER_FRAME,
+      (int)printingHelper->maxAddrLen, BACTRACE_PANEL_HEADER_ADDRESS,
+      (int)maxFunctionNameLength, BACTRACE_PANEL_HEADER_NAME,
+      (int)printingHelper->maxObjPathLen, BACTRACE_PANEL_HEADER_PATH
+   );
+
+   RichString_appendnAscii(out, CRT_colors[BACKTRACE_HEADER], line, len);
+   free(line);
+}
+
+static void BacktraceFrame_highlightBasename(const BacktraceFrame* frame, RichString* out, char* line, int objectPathStart) {
+   assert(objectPathStart >= 0);
+
+   char* basename = frame->backtracePanel->process->procExe + frame->backtracePanel->process->procExeBasenameOffset;
+   char* object = line + objectPathStart;
+   if (!basename || !object) {
```

Comment:
Looks like something is wrong with this conditional. Perhaps you mean `(!frame->backtracePanel->process->procExe || !line)` because you should ignore the character offset when the pointer is null already.

Reply:
Good catch, but also not quite correct, because `object` is a pointer addition too and thus should check `line`.

Alternative:
```c
   char* basename = frame->backtracePanel->process->procExe ? frame->backtracePanel->process->procExe + frame->backtracePanel->process->procExeBasenameOffset : NULL;
   char* object = line ? line + objectPathStart : NULL;
   if (!basename || !object) {
```
In which case the current check would work just fine.

You could even do a small macro `#define PTR_ADD_OFFSET(ptr, offset) (ptr) ? ((ptr) + (offset)) : NULL`.

Reply:
I assumed if `procExe` is null, then `procExeBasenameOffset` equals 0. But it may be false.
Good catch! Thank you!